File: structural-design-pattern/bridge-pattern/bridge.go
// Bridge Pattern
// ==============
// Category: Structural Design Pattern
//
// Intent:
//   Decouple an abstraction from its implementation so both can vary
//   independently. Avoids N × M combinatorial explosion.
//
// When to use:
//   - When you'd need many class combinations without the bridge
//   - When both abstraction and implementation need to be extensible
//   - When you want to swap implementations at runtime
//
// Key Participants:
//   - Abstraction (struct + interface): High-level logic
//   - Implementation (interface): Low-level mechanism
//   - Bridge: The implementation interface field in the abstraction struct

package main

import "fmt"

// ---------------------------------------------------------------------------
// Implementation interface — delivery channels
// ---------------------------------------------------------------------------

type MessageChannel interface {
	Send(title, body, recipient string) string
}

type EmailChannel struct{}

func (e *EmailChannel) Send(title, body, recipient string) string {
	return fmt.Sprintf("[Email → %s] Subject: %s | %s", recipient, title, body)
}

type SMSChannel struct{}

func (s *SMSChannel) Send(title, body, recipient string) string {
	short := body
	if len(body) > 160 {
		short = body[:160]
	}
	return fmt.Sprintf("[SMS → %s] %s", recipient, short)
}

type SlackChannel struct {
	Workspace string
}

func (s *SlackChannel) Send(title, body, recipient string) string {
	return fmt.Sprintf("[Slack#%s → @%s] *%s*\n%s", s.Workspace, recipient, title, body)
}

// ---------------------------------------------------------------------------
// Abstraction interface — notification types
// ---------------------------------------------------------------------------

type Notification interface {
	Notify(recipient, message string) string
}

// RegularNotification — standard priority
type RegularNotification struct {
	Channel MessageChannel // The "bridge" — composition
}

func (n *RegularNotification) Notify(recipient, message string) string {
	return n.Channel.Send("Notification", message, recipient)
}

// UrgentNotification — high priority with urgency markers
type UrgentNotification struct {
	Channel MessageChannel
}

func (n *UrgentNotification) Notify(recipient, message string) string {
	body := fmt.Sprintf("%s — Immediate action required!", message)
	return n.Channel.Send("URGENT", body, recipient)
}

// ScheduledNotification — deferred delivery
type ScheduledNotification struct {
	Channel       MessageChannel
	ScheduledTime string
}

func (n *ScheduledNotification) Notify(recipient, message string) string {
	title := fmt.Sprintf("Scheduled (%s)", n.ScheduledTime)
	body := fmt.Sprintf("%s [Will be delivered at %s]", message, n.ScheduledTime)
	return n.Channel.Send(title, body, recipient)
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
func main() {
	// Without Bridge: 3 types × 3 channels = 9 structs
	// With Bridge:    3 types + 3 channels  = 6 structs

	fmt.Println("=== Regular via Email ===")
	fmt.Println((&RegularNotification{Channel: &EmailChannel{}}).Notify(
		"alice@co.com", "Your report is ready"))

	fmt.Println("\n=== Urgent via SMS ===")
	fmt.Println((&UrgentNotification{Channel: &SMSChannel{}}).Notify(
		"555-1234", "Server CPU at 98%"))

	fmt.Println("\n=== Urgent via Slack ===")
	fmt.Println((&UrgentNotification{Channel: &SlackChannel{Workspace: "ops"}}).Notify(
		"oncall", "Database failover triggered"))

	fmt.Println("\n=== Scheduled via Email ===")
	fmt.Println((&ScheduledNotification{
		Channel:       &EmailChannel{},
		ScheduledTime: "2025-03-15 09:00",
	}).Notify("team@co.com", "Weekly standup reminder"))

	// Swap channel at runtime
	fmt.Println("\n=== Same notification, different channels ===")
	channels := []MessageChannel{
		&EmailChannel{},
		&SMSChannel{},
		&SlackChannel{Workspace: "general"},
	}
	for _, ch := range channels {
		fmt.Println((&UrgentNotification{Channel: ch}).Notify("bob", "Deploy failed"))
	}
}

File: structural-design-pattern/bridge-pattern/bridge.rs
//! Bridge Pattern
//! ==============
//! Category: Structural Design Pattern
//!
//! Intent:
//!   Decouple an abstraction from its implementation so both can vary
//!   independently. Avoids combinatorial explosion of types.
//!
//! When to use:
//!   - When you'd need N × M structs without the bridge
//!   - When both abstraction and implementation need to be extensible
//!
//! Key Participants:
//!   - Abstraction (trait): High-level logic (notification urgency)
//!   - Implementation (trait): Low-level mechanism (delivery channel)
//!   - Bridge: The Box<dyn Implementation> held by the abstraction

// ---------------------------------------------------------------------------
// Implementation trait — delivery channels
// ---------------------------------------------------------------------------
trait MessageChannel {
    fn send(&self, title: &str, body: &str, recipient: &str) -> String;
}

struct EmailChannel;
impl MessageChannel for EmailChannel {
    fn send(&self, title: &str, body: &str, recipient: &str) -> String {
        format!("[Email → {}] Subject: {} | {}", recipient, title, body)
    }
}

struct SMSChannel;
impl MessageChannel for SMSChannel {
    fn send(&self, title: &str, body: &str, recipient: &str) -> String {
        let short = if body.len() > 160 { &body[..160] } else { body };
        format!("[SMS → {}] {}", recipient, short)
    }
}

struct SlackChannel {
    workspace: String,
}
impl SlackChannel {
    fn new(workspace: &str) -> Self {
        Self { workspace: workspace.to_string() }
    }
}
impl MessageChannel for SlackChannel {
    fn send(&self, title: &str, body: &str, recipient: &str) -> String {
        format!(
            "[Slack#{} → @{}] *{}*\n{}",
            self.workspace, recipient, title, body
        )
    }
}

// ---------------------------------------------------------------------------
// Abstraction — notification types hold a reference to a channel (the bridge)
// ---------------------------------------------------------------------------
trait Notification {
    fn notify(&self, recipient: &str, message: &str) -> String;
}

/// Regular priority — sends message as-is.
struct RegularNotification {
    channel: Box<dyn MessageChannel>,
}
impl Notification for RegularNotification {
    fn notify(&self, recipient: &str, message: &str) -> String {
        self.channel.send("Notification", message, recipient)
    }
}

/// Urgent — adds urgency markers.
struct UrgentNotification {
    channel: Box<dyn MessageChannel>,
}
impl Notification for UrgentNotification {
    fn notify(&self, recipient: &str, message: &str) -> String {
        let body = format!("{} — Immediate action required!", message);
        self.channel.send("URGENT", &body, recipient)
    }
}

/// Scheduled — includes delivery time.
struct ScheduledNotification {
    channel: Box<dyn MessageChannel>,
    scheduled_time: String,
}
impl Notification for ScheduledNotification {
    fn notify(&self, recipient: &str, message: &str) -> String {
        let title = format!("Scheduled ({})", self.scheduled_time);
        let body = format!(
            "{} [Will be delivered at {}]",
            message, self.scheduled_time
        );
        self.channel.send(&title, &body, recipient)
    }
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
fn main() {
    // Without Bridge: 3 types × 3 channels = 9 structs
    // With Bridge:    3 types + 3 channels  = 6 structs

    println!("=== Regular via Email ===");
    let notif = RegularNotification {
        channel: Box::new(EmailChannel),
    };
    println!("{}", notif.notify("alice@co.com", "Your report is ready"));

    println!("\n=== Urgent via SMS ===");
    let notif = UrgentNotification {
        channel: Box::new(SMSChannel),
    };
    println!("{}", notif.notify("555-1234", "Server CPU at 98%"));

    println!("\n=== Urgent via Slack ===");
    let notif = UrgentNotification {
        channel: Box::new(SlackChannel::new("ops")),
    };
    println!("{}", notif.notify("oncall", "Database failover triggered"));

    println!("\n=== Scheduled via Email ===");
    let notif = ScheduledNotification {
        channel: Box::new(EmailChannel),
        scheduled_time: "2025-03-15 09:00".to_string(),
    };
    println!("{}", notif.notify("team@co.com", "Weekly standup reminder"));

    // Swap channel at runtime
    println!("\n=== Same notification type, different channels ===");
    let channels: Vec<Box<dyn MessageChannel>> = vec![
        Box::new(EmailChannel),
        Box::new(SMSChannel),
        Box::new(SlackChannel::new("general")),
    ];
    for ch in channels {
        let notif = UrgentNotification { channel: ch };
        println!("{}", notif.notify("bob", "Deploy failed"));
    }
}
